src/Keboola/DbExtractor/Retrier.php

```diff
@@ -0,0 +1,63 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Keboola\DbExtractor;
+
+use Keboola\DbExtractor\Exception\UserException;
+
+class Retrier
+{
+    /**
+     * @var  int
+     */
+    private $maxTries;
+
+    /**
+     * @var array
+     */
+    private $expectedExceptions;
+
+    /**
+     * @var Logger
+     */
+    private $logger;
+
+    public function __construct(int $maxTries, array $expectedExceptions, Logger $logger)
+    {
+        $this->maxTries = $maxTries;
+        $this->expectedExceptions = $expectedExceptions;
+        $this->logger = $logger;
+    }
+
+    /**
+     * @param callable $method
+     * @param array    $methodParams
+     * @return mixed
+     * @throws UserException
+     */
+    public function retry(
+        callable $method,
+        array $methodParams
+    ) {
+        $counter = 0;
+
+        while (true) {
+            try {
+                return call_user_func($method, $methodParams);
+            } catch (\Throwable $e) {
+                foreach ($this->expectedExceptions as $expectedException) {
+                    if ($e instanceof $expectedException) {
+                        $counter++;
+                        if ($counter > $this->maxTries) {
+                            throw $e;
+                        }
+                        $this->logger->info(sprintf('%s. Retrying... [%dx]', $e->getMessage(), $counter));
+                    } else {
+                        throw $e;
+                    }
+                }
+            }
+        }
+    }
```

Comment:
I'm wondering whether it would be worth to extend this https://github.com/vkartaviy/retry/blob/master/src/Retry/RetryProxy.php

i see 
pros:
- exponential backoff and other ready-to-use strategies
(the above needs some sleep definitely)
cons:
- impossible to pass $logger in ctor

perhaps using only the interfaces might be a solution? not sure - thinking aloud

Reply:
how about extending the original library? the author may be open to pull requests.

Reply:
if i were him, i wouldn't accept it :)

Reply:
Yeah, I had thought about implementing that interface in a very similar manner to https://github.com/vkartaviy/retry/blob/master/src/Retry/RetryProxy.php only with the addition of the logging functionality.  I'll give that a go.  but first I'm going to separate this beast into another PR.  cheers guys.